Implement flat kubemodel architecture with bingen serialization and metric hydration

spa-raj opened the pull request:
## What does this PR change?
*  Implements a new flat kubemodel architecture with bingen annotations for binary serialization and full metric hydration from costmodel datasource, including registration functions and compute methods for all Kubernetes resource types (nodes, pods, containers, controllers, services, PVCs, volumes)

## Does this PR relate to any other PRs?
* This PR is built upon the work from #3435 

## How will this PR impact users?
*  No immediate user-facing impact. This PR lays the foundation for Data Model 2.0 by enabling efficient binary serialization and structured metric collection, which will support future features like improved storage and analytics capabilities.

## Does this PR address any GitHub or Zendesk issues?
* Closes ...

## How was this PR tested?
*  Verified through successful compilation of all kubemodel and costmodel packages after fixing result struct field mappings. Part of ongoing LFX mentorship project testing will include integration testing with the broader Data Model 2.0 pipeline

## Does this PR require changes to documentation?
* No

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next OpenCost release? If not, why not?
* This PR contains breaking changes and I leave it to the maintainers to decide its release.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Adds a new flat kubemodel with bingen-annotated schemas, export pipeline and pathing, and costmodel computation/hydration, plus node public IP helpers and pathing interface updates.
> 
> - **KubeModel Data Model (new)**:
>   - Introduces flat `kubemodel` package with bingen-annotated types (`Cluster`, `Namespace`, `Node`, `Pod`, `Container`, `Service`, `Volume`, `PVC`, `Controller`, `ResourceQuota`, GPU structures, stats/units, `KubeModelSet`).
>   - Adds validation/clone utilities and tests.
> - **CostModel Integration**:
>   - Implements `ComputeKubeModel` and helpers to hydrate all resources from metrics; computes `PublicIPSeconds` per node via cluster cache.
> - **Exporter/Pipelines**:
>   - Adds KubeModel pipeline source (`opencost/exporter/kubemodel`) and wires it into `PipelineExportControllers` with start/stop and default resolutions.
>   - Extends mocks/tests to cover KubeModel exporting.
> - **Pathing**:
>   - Adds KubeModel-specific storage formatter (`kubemodelpath.go`) and routes KubeModel pipeline to FinOps-agent root pathing.
>   - Simplifies `StoragePathFormatter` (removes `RootDir()`); adjusts bingen/event pathing and constructors.
> - **Cluster Cache**:
>   - Adds `Node.GetPublicIPAddresses()` and `Node.GetPublicIPCount()` helpers.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit f0b8bec6981b4f62f88bfa96d7bccd5ddf397b99. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

spa-raj committed:
Update kubemodel with MVP device architecture

   - Replace GPU-specific arrays with generic map-based device model
   - Add Device, DeviceType, and DeviceUsage structs
   - Update Container to use Devices map[string]*DeviceUsage
   - Update Container time fields: CreationTime/DeletionTime -> Start/End
   - Update KubeModelSet to use Devices map[string]*Device
   - Remove old gpu_device.go and gpu_usage.go files
   - Update CPU/RAM/Storage units to seconds and uint64 for consistency
   - Focus on OpenCost-compatible metrics for MVP

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add remaining kubemodel Go structs from feature/kubemodel-go-structs

   Complete the kubemodel package by adding all remaining Go struct files:
   - cluster.go
   - controller.go
   - diagnostic.go
   - namespace.go
   - node.go
   - provider.go
   - pvc.go
   - resolution.go
   - resource.go
   - resourcequota.go
   - service.go
   - volume.go
   - window.go

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Implement fully hierarchical kubemodel structure

Restructure KubeModelSet to use nested hierarchy instead of flat maps,
following the architectural pattern from PR #3435. This creates a more
intuitive and maintainable model that mirrors Kubernetes' actual resource
relationships.

Hierarchical structure:
 - Cluster
   - Nodes → Pods → Containers → Volumes/Devices
   - Namespaces → Controllers/ResourceQuotas
   - PersistentVolumes
   - LoadBalancers

Key changes:
 - Simplify KubeModelSet to contain only Metadata, Window, Cluster, and indexes
 - Nest Nodes, Namespaces, PersistentVolumes, LoadBalancers in Cluster
 - Nest Pods and EphemeralVolumes in Node
 - Nest Containers, AttachedDevices, PVCs in Pod
 - Nest Volumes and Devices in Container
 - Nest Controllers and ResourceQuotas in Namespace
 - Update all resources to use Start/End time.Time fields for consistency
 - Update KubeModelSetMetadata with CompletedAt field
 - Implement RegisterNode and GetNodeByName methods
 - Update RegisterNamespace to work with nested hierarchy
 - Maintain internal indexes for O(1) name-based lookups
 - Update IsEmpty to check nested collections

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add bingen annotations and Stats support to hierarchical kubemodel

  Implement complete @bingen:field annotations for binary codec generation, add Stats type for multi-dimensional resource tracking, and enhance kubemodel.go with O(1) index-based
  registration and lookup methods.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Integrate KubeModel pipeline with Stats API updates and incremental resource building

  Update Stats constants, modify ResourceQuantities.Set() for incremental stat values, add KubeModel export pipeline with controllers/mocks, and implement
  CostModel.ComputeKubeModel() for namespace computation.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Implement cluster ID computation and complete resource quota integration for hierarchical kubemodel

Uses kube-system namespace UID for cluster ID uniqueness, adds diagnostic error handling, and implements full resource quota computation with all spec/status metrics properly nested in parent namespaces.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Updated kubemodel from heirarchical to flat

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add bingen annotations to kubemodel structs and standardize time fields

  Replace CreationTime/DeletionTime with Start/End across all resources, add @bingen:generate annotations
  with version=1 fields for binary serialization. Add metadata.go, stats.go, stats_test.go, and unit.go.
  Remove unused resolution.go.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add resource registration functions and fix costmodel integration

  Implement Register* functions for Pod, Node, Controller, Service, PVC, Volume, Container, and Device in KubeModelSet.
  Fix costmodel to use correct field names (UID, flat namespace structure) and Stat constants (StatAvg/StatMax).

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add metric hydration for all kubemodel resource types

  Implement compute functions to populate KubeModelSet with metrics from datasource for nodes, pods, containers, controllers, services, PVCs, and volumes. Fix result struct field
  name mismatches (PodUID->UID, PV->PersistentVolume, etc.) to match decoder definitions.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Refactor storage path formatter to support KubeModel pipeline routing

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj commented:
@Sean-Holcomb @nikovacevic @ameijer  I have reverted to a flat model and added more implementations of hydration of the metrics as per Niko's PR.

ameijer commented on core/pkg/model/kubemodel/kubemodel.go:
```diff
@@ -0,0 +1,274 @@
+	PersistentVolumeClaims map[string]*PersistentVolumeClaim `json:"pvcs,omitempty"`       // @bingen:field[version=1]
+	Services               map[string]*Service               `json:"services,omitempty"`   // @bingen:field[version=1]
+	Volumes                map[string]*Volume                `json:"volumes,omitempty"`    // @bingen:field[version=1]
+	idx                    *kubeModelSetIndexes              // @bingen:field[ignore]
```
shall we consider removing the indexes?

ameijer replied:
edit - not a concern, we can keep these

spa-raj committed:
Refactor KubeModel storage path to use date-based hierarchy and pre-computed paths

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

Sean-Holcomb commented on core/pkg/model/kubemodel/window.go:
```diff
@@ -0,0 +1,9 @@
+// @bingen:generate:Window
+type Window struct {
+	Start time.Time `json:"start"` // @bingen:field[version=1]
+	End   time.Time `json:"end"`   // @bingen:field[version=1]
```
I prefer this to opencost.Window but are we moving to this rather than start and resolution?

spa-raj replied:
During our previous discussion, Jay recommended using dynamic windows as an approach, and Niko expressed agreement with this direction. 

I suggest we revisit this topic in our next standup meeting to align on the implementation details.

aaliomer-ibm replied:
I think the discussion was about the codebase already using resolutions and the huge amount of work it would require to refactor if i remember correctly. Not too familiar with the codebase to give a level of effort estimate

ameijer replied:
benefit of Start + resolution is that we can have constraints around allowed resolutions

ameijer replied:
we will keep it start and end

Sean-Holcomb reviewed (commented):
Looks really good, great work.

ameijer commented on core/pkg/model/kubemodel/service.go:
```diff
@@ -0,0 +1,29 @@
+
+// @bingen:generate:Service
+type Service struct {
+	ID                   string            `json:"id"`                   // @bingen:field[version=1]
```
Load balancer vs service
ID vs UID

aaliomer-ibm replied:
unrelated but I'd add every k8s object has a UUID id type. everything else should use a string id (container id should be a struct with pod-<name/id>, container-name)

spa-raj replied:
@Sean-Holcomb @nikovacevic Do we want ID for containers? We don't have it in our current model. Shall I update the data type for all the K8s object UIDs to UUID?

Sean-Holcomb replied:
The map in the flat struct should be keyed on a concatenated string of the pod UID and the container name. It should be a string because keying on a struct limits serialization options. I don't think this key id needs to be part of the container struct since it is already available in other fields of the container struct.

aaliomer-ibm commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,25 @@
+	End                        time.Time               `json:"end"`                        // @bingen:field[version=1]
+	CpuMillicoreSeconds        uint64                  `json:"cpuMillicoreSeconds"`        // @bingen:field[version=1]
+	CpuMillicoreRequestAverage uint64                  `json:"cpuMillicoreRequestAverage"` // @bingen:field[version=1]
+	CpuMillicoreUsageAverage   uint64                  `json:"cpuMillicoreUsageAverage"`   // @bingen:field[version=1]
```
nit: Nano-core is the smallest unit. dont think it is a big deal however

aaliomer-ibm commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,31 @@
+	StorageByteSeconds         uint64                            `json:"storageByteSeconds"`               // @bingen:field[version=1]
+	NetworkTransferBytes       uint64                            `json:"networkTransferBytes"`             // @bingen:field[version=1]
+	NetworkReceiveBytes        uint64                            `json:"networkReceiveBytes"`              // @bingen:field[version=1]
+	Containers                 map[string]*Container             `json:"containers"`                       // @bingen:field[version=1]
```
I dont think this should be here. i would not have references to pointers. instead, i'd have a set of device IDs / volume claims.
Wasnt the index object supposed to do this role? this is no longer flat. but now is a graph which i dont think it is the intention

spa-raj committed:
Refactor device model to GPU-specific types

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Updated the kubemodel structs to have UIDs consistently

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Added unit tests for the kubemodel

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Updated kubemodel to use UIDs instead of IDs

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Standardize kubemodel field naming and units for consistency

  Convert to RAM (all caps), CPU millicores, time in seconds, GPU uint64, and fix Byte/Bytes naming inconsistencies across Container, Pod, Node, GPU structs.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -0,0 +1,13 @@
+	Account  string    `json:"account"`  // @bingen:field[version=1]
+	Name     string    `json:"name"`     // @bingen:field[version=1]
+	Start    time.Time `json:"start"`    // @bingen:field[version=1]
+	End      time.Time `json:"end"`      // @bingen:field[version=1]
```
were we going to start going with Start + resolution?

ameijer replied:
looking back over prior reviews, where did we land with https://github.com/opencost/opencost/pull/3443#discussion_r2517618073?

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	RAMBytesRequestAverage     uint64                     `json:"ramBytesRequestAverage"`     // @bingen:field[version=1]
+	RAMBytesUsageAverage       uint64                     `json:"ramBytesUsageAverage"`       // @bingen:field[version=1]
+	RAMBytesUsageMax           uint64                     `json:"ramBytesUsageMax"`           // @bingen:field[version=1]
+	StorageBytesSeconds        uint64            `json:"storageBytesSeconds"`        // @bingen:field[version=1]
```
nit- fix spacing

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	CpuMillicoreRequestAverage uint64                     `json:"cpuMillicoreRequestAverage"` // @bingen:field[version=1]
+	CpuMillicoreUsageAverage   uint64                     `json:"cpuMillicoreUsageAverage"`   // @bingen:field[version=1]
+	CpuMillicoreUsageMax       uint64                     `json:"cpuMillicoreUsageMax"`       // @bingen:field[version=1]
+	RAMBytesSeconds            uint64                     `json:"ramBytesSeconds"`            // @bingen:field[version=1]
```
nit - I think all these should be byte-seconds, not bytes-seconds

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	StorageBytesSeconds        uint64            `json:"storageBytesSeconds"`        // @bingen:field[version=1]
+	StorageBytesRequestAverage uint64            `json:"storageBytesRequestAverage"` // @bingen:field[version=1]
+	StorageBytesUsageAverage   uint64            `json:"storageBytesUsageAverage"`   // @bingen:field[version=1]
+	StorageBytesUsageMax       uint64            `json:"storageBytesUsageMax"`       // @bingen:field[version=1]
```
I wonder if we need support for a few kinds of storage:

1. ephemeral
2. PV

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	StorageBytesRequestAverage uint64            `json:"storageBytesRequestAverage"` // @bingen:field[version=1]
+	StorageBytesUsageAverage   uint64            `json:"storageBytesUsageAverage"`   // @bingen:field[version=1]
+	StorageBytesUsageMax       uint64            `json:"storageBytesUsageMax"`       // @bingen:field[version=1]
+	Diagnostic                 *DiagnosticResult `json:"diagnostic,omitempty"`       // @bingen:field[version=1]
```
what was the consensus on where diagnosticResult was going?

ameijer commented on core/pkg/model/kubemodel/gpudevice.go:
```diff
@@ -0,0 +1,77 @@
+// GPUDevice represents a GPU device with DCGM integration (provisioned resource)
+// This tracks available GPU capacity on a node
+type GPUDevice struct {
+	UID                string            `json:"uid"`                      // @bingen:field[version=1] GPU UUID (hardware identifier)
```
fix spacing

ameijer commented on core/pkg/model/kubemodel/gpudevice.go:
```diff
@@ -0,0 +1,77 @@
+	GpuRequestPercentageAverage float64           `json:"gpuRequestPercentageAverage"` // @bingen:field[version=1] GPU request average percentage (0-100)
+	GpuUsagePercentageAverage   float64           `json:"gpuUsagePercentageAverage"`   // @bingen:field[version=1] GPU usage average percentage (0-100)
+	GpuUsagePercentageMax       float64           `json:"gpuUsagePercentageMax"`       // @bingen:field[version=1] GPU usage max percentage (0-100)
+	MemoryBytes                 uint64            `json:"memoryBytes"`                 // @bingen:field[version=1] GPU memory capacity in bytes
```
did we want the device tree to hold pricing information too?

ameijer replied:
byte-seconds?

ameijer replied:
no pricing info in kube model

ameijer commented on core/pkg/model/kubemodel/gpuusage.go:
```diff
@@ -0,0 +1,66 @@
+	GpuRequestPercentageAverage float64           `json:"gpuRequestPercentageAverage"` // @bingen:field[version=1] GPU request in percentage (0-100)
+	GpuUsagePercentageAverage   float64           `json:"gpuUsagePercentageAverage"`   // @bingen:field[version=1] GPU usage average percentage (0-100)
+	GpuUsagePercentageMax       float64           `json:"gpuUsagePercentageMax"`       // @bingen:field[version=1] GPU usage max percentage (0-100)
+	MemoryBytesUsed             uint64            `json:"memoryBytesUsed"`             // @bingen:field[version=1] GPU memory usage in bytes
```
byte-seconds?

aaliomer-ibm replied:
+1

ameijer commented on core/pkg/model/kubemodel/kubemodel.go:
```diff
@@ -0,0 +1,296 @@
+
+// @bingen:generate[stringtable]:KubeModelSet
+type KubeModelSet struct {
+	Metadata               *Metadata                         `json:"meta"`                  // @bingen:field[version=1]
```
so should we remove the diagnostic results from each struct then ?

Sean-Holcomb replied:
That seems right to me

ameijer commented on core/pkg/model/kubemodel/namespace.go:
```diff
@@ -0,0 +1,17 @@
+	Start           time.Time                 `json:"start"`                    // @bingen:field[version=1]
+	End             time.Time                 `json:"end"`                      // @bingen:field[version=1]
+	Controllers     map[string]*Controller    `json:"controllers,omitempty"`    // @bingen:field[ignore]
+	ResourceQuotas  map[string]*ResourceQuota `json:"resourceQuotas,omitempty"` // @bingen:field[ignore]
```
I think these can be removed - these objects have uid pointers back to the namespace

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	RAMBytes                  uint64             `json:"ramBytes"`                   // @bingen:field[version=1]
+	CpuCost                   float64            `json:"cpuCost"`                    // @bingen:field[version=1]
+	RAMCost                   float64            `json:"ramCost"`                    // @bingen:field[version=1]
+	GpuCost                   float64            `json:"gpuCost"`                    // @bingen:field[version=1]
```
should these be price rather than cost?

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	Annotations          map[string]string  `json:"annotations,omitempty"`      // @bingen:field[version=1]
+	Start                     time.Time          `json:"start"`                      // @bingen:field[version=1]
+	End                       time.Time          `json:"end"`                        // @bingen:field[version=1]
+	CpuMillicores             uint64             `json:"cpuMillicores"`              // @bingen:field[version=1]
```
millicore-seconds?

aaliomer-ibm replied:
+1

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	Start                     time.Time          `json:"start"`                      // @bingen:field[version=1]
+	End                       time.Time          `json:"end"`                        // @bingen:field[version=1]
+	CpuMillicores             uint64             `json:"cpuMillicores"`              // @bingen:field[version=1]
+	RAMBytes                  uint64             `json:"ramBytes"`                   // @bingen:field[version=1]
```
byte-seconds?

aaliomer-ibm replied:
+1 always use commutative units

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	RAMBytesUsageAverage      uint64             `json:"ramBytesUsageAverage"`       // @bingen:field[version=1]
+	RAMBytesUsageMax          uint64             `json:"ramBytesUsageMax"`           // @bingen:field[version=1]
+	Pods                 map[string]*Pod    `json:"pods"`                       // @bingen:field[version=1]
+	EphemeralVolumes     map[string]*Volume `json:"ephemeralVolumes,omitempty"` // @bingen:field[version=1]
```
won't these objects have pointers back to this node? so we can toss this mapping?

aaliomer-ibm replied:
can a volume be de-attached and attached to a different node? how will this impact this model? then you will have multiple nodes pointing to the same volume. For example, a stateful set will follow the volume not the node necessarily i think but not sure

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,31 @@
+	RAMBytesRequestAverage     uint64                            `json:"ramBytesRequestAverage"`           // @bingen:field[version=1]
+	RAMBytesUsageAverage       uint64                            `json:"ramBytesUsageAverage"`             // @bingen:field[version=1]
+	RAMBytesUsageMax           uint64                            `json:"ramBytesUsageMax"`                 // @bingen:field[version=1]
+	StorageBytesSeconds        uint64                            `json:"storageBytesSeconds"`              // @bingen:field[version=1]
```
break this up into two kinds?

ameijer replied:
actually - do we even need this? can't this be dynamically calculated?

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,31 @@
+	RAMBytesSeconds            uint64                            `json:"ramBytesSeconds"`                  // @bingen:field[version=1]
+	RAMBytesRequestAverage     uint64                            `json:"ramBytesRequestAverage"`           // @bingen:field[version=1]
+	RAMBytesUsageAverage       uint64                            `json:"ramBytesUsageAverage"`             // @bingen:field[version=1]
+	RAMBytesUsageMax           uint64                            `json:"ramBytesUsageMax"`                 // @bingen:field[version=1]
```
do we need these usages on the pod? Can't these be dynamically calculated?

aaliomer-ibm replied:
you can get an average from byteseconds but not max/min/p95 etc...

spa-raj replied:
I have removed averages and kept only max

ameijer commented on core/pkg/model/kubemodel/resource.go:
```diff
@@ -0,0 +1,33 @@
+	ResourceCPU     Resource = "cpu"
+	ResourceMemory  Resource = "memory"
+	ResourceGPU     Resource = "gpu"
+	ResourceStorage Resource = "storage"
```
is this dead?>

Sean-Holcomb replied:
I don't believe it is, but we also need to pick if we are going to use this or not and use it everywhere consistently

ameijer commented on core/pkg/opencost/mock.go:
```diff
@@ -1013,3 +1015,10 @@ func GenerateMockCloudCostSet(start, end time.Time, provider, integration string
+
+// GenerateMockKubeModelSet creates generic allocation set without idle allocations
+func GenerateMockKubeModelSet(start time.Time) *kubemodel.KubeModelSet {
+	// TODO: fill out mocks
```
were we planning on populating this in scope of this PR?

ameijer commented on core/pkg/model/kubemodel/volume.go:
```diff
@@ -0,0 +1,19 @@
+import "time"
+
+// @bingen:generate:Volume
+type Volume struct {
```
should we add an 'ephemeral volume' type as well? Jay what kinds of things do we want to track for ephemeral storage?

ameijer replied:
at min, pointer to node

aaliomer-ibm replied:
in cldy, we have two types of storage. Root storage (storage attached to the node lifecycle aka ephemeral) and attached storage.
Sometimes in AWS, Node storage is backed by an EBS volume. 
The attached root storage is billed separately by AWS for example and an EBS. We do keep the ID of it to later allocate the cost to it. 

Getting back at your point whether both should be separate, it appears so. Root volume should be a part of the node instead of the volume IMO

spa-raj replied:
Perhaps we can rename "volume" to "PersistantVolume" and add these to node?

`EphemeralStorageBytes       uint64 'json:"ephemeralStorageBytes"'`
`EphemeralStorageByteSeconds uint64 'json:"ephemeralStorageByteSeconds"'`
`RootVolumeProviderID        string 'json:"rootVolumeProviderId,omitempty"'`

ameijer replied:
I think for root volume, we do need to record its total size in the Node object sparsh. For the consumers, I think at container level we enhance storage to be a map which can have different types, PVC vs ephemeral

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+import "time"
+
+// @bingen:generate:Node
+type Node struct {
```
can we also add public IPs as a cost incurring resource type to this node?

ameijer commented on core/pkg/model/kubemodel/service.go:
```diff
@@ -0,0 +1,29 @@
+}
+
+// @bingen:generate:Service
+type Service struct {
```
I think we also need a corresponding loadBalancer type

ameijer replied:
tabling this, just service will work

ameijer commented on core/pkg/model/kubemodel/cluster.go:
```diff
@@ -0,0 +1,13 @@
+import "time"
+
+// @bingen:generate:Cluster
+type Cluster struct {
```
should cluster have a price field as well to capture mgmt costs?

ameijer replied:
no pricing in kube model

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	CpuMillicores             uint64             `json:"cpuMillicores"`              // @bingen:field[version=1]
+	RAMBytes                  uint64             `json:"ramBytes"`                   // @bingen:field[version=1]
+	CpuCost                   float64            `json:"cpuCost"`                    // @bingen:field[version=1]
+	RAMCost                   float64            `json:"ramCost"`                    // @bingen:field[version=1]
```
remove all pricing

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,31 @@
+	NetworkReceiveBytes        uint64                            `json:"networkReceiveBytes"`              // @bingen:field[version=1]
+	Containers                 map[string]*Container             `json:"containers"`                       // @bingen:field[version=1]
+	AttachedGPUDevices         map[string]*GPUDevice             `json:"attachedGpuDevices,omitempty"`     // @bingen:field[version=1]
+	PersistentVolumeClaims     map[string]*PersistentVolumeClaim `json:"persistentVolumeClaims,omitempty"` // @bingen:field[version=1]
```
remove these since they will have pointers to this?

aaliomer-ibm commented on core/pkg/model/kubemodel/gpuusage.go:
```diff
@@ -0,0 +1,66 @@
+	GpuDeviceUID                string            `json:"gpuDeviceUid"`                // @bingen:field[version=1] Reference to the GPU device being used
+	GpuSeconds                  float64           `json:"gpuSeconds"`                  // @bingen:field[version=1] GPU usage in device-seconds consumed
+	GpuRequestPercentageAverage float64           `json:"gpuRequestPercentageAverage"` // @bingen:field[version=1] GPU request in percentage (0-100)
+	GpuUsagePercentageAverage   float64           `json:"gpuUsagePercentageAverage"`   // @bingen:field[version=1] GPU usage average percentage (0-100)
```
I understand that GPU is like CPU, it has memory and compute dimensions. Both need to be exposed. Power consumption is also a metric that we probably want to keep track of

spa-raj replied:
I have added fields for power consumption but I think opencost is not currently scraping DCGM metrics for the same. We need to update our scraper as well to see power metrics in effect.

aaliomer-ibm commented on core/pkg/model/kubemodel/namespace.go:
```diff
@@ -0,0 +1,17 @@
+	Annotations     map[string]string         `json:"annotations"`              // @bingen:field[version=1]
+	Start           time.Time                 `json:"start"`                    // @bingen:field[version=1]
+	End             time.Time                 `json:"end"`                      // @bingen:field[version=1]
+	Controllers     map[string]*Controller    `json:"controllers,omitempty"`    // @bingen:field[ignore]
```
I thought controllers and ResourceQuotas are going to be flat

aaliomer-ibm commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,27 @@
+	Name                 string             `json:"name"`                       // @bingen:field[version=1]
+	Labels               map[string]string  `json:"labels,omitempty"`           // @bingen:field[version=1]
+	Annotations          map[string]string  `json:"annotations,omitempty"`      // @bingen:field[version=1]
+	Start                     time.Time          `json:"start"`                      // @bingen:field[version=1]
```
whats the start and end are for?

spa-raj replied:
Specifically for Nodes - they can join/leave clusters dynamically (auto-scaling, failures, maintenance).
Also, cumulative metrics are meaningless without a time range.

aaliomer-ibm commented on core/pkg/model/kubemodel/unit.go:
```diff
@@ -0,0 +1,20 @@
+	UnitMi            = "Mi"
+	UnitGB            = "GB"
+	UnitGPU           = "GPU"
+	UnitSecond        = "s"
```
not needed anymore

spa-raj committed:
Refactor kubemodel to use singular Byte and cumulative Seconds units

  Standardize field naming across all kubemodel structs: use singular Byte instead of Bytes, add Seconds suffix for cumulative time-based metrics, remove cost fields, and add
  GPU power metrics for future implementation.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Removes PV pricing queries and cost calculation from kmComputeVolumes, focusing kubemodel on resource metrics rather than pricing concerns

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Add public IP tracking to kubemodel Node for cost attribution

- Adds PublicIPSeconds field to kubemodel.Node to track public IP addresses
  as a cost-incurring resource. The field represents cumulative public IP
  allocation (count × seconds) calculated from Kubernetes node ExternalIP
  addresses.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer commented on core/pkg/clustercache/clustercache.go:
```diff
@@ -171,6 +171,28 @@ type ResourceQuota struct {
+	var publicIPs []string
+	for _, addr := range n.Status.Addresses {
+		if addr.Type == v1.NodeExternalIP {
+			publicIPs = append(publicIPs, addr.Address)
```
wow that is nice that we have this so accessible.

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	StorageByteRequestAverage  uint64            `json:"storageByteRequestAverage"`  // @bingen:field[version=1]
+	StorageByteUsageAverage    uint64            `json:"storageByteUsageAverage"`    // @bingen:field[version=1]
+	StorageByteUsageMax        uint64            `json:"storageByteUsageMax"`        // @bingen:field[version=1]
+	Diagnostic                 *DiagnosticResult `json:"diagnostic,omitempty"`       // @bingen:field[version=1]
```
following up here - I think we only want Diagnostics in the root kubemodel object

ameijer commented on core/pkg/model/kubemodel/gpuusage.go:
```diff
@@ -0,0 +1,92 @@
+	MemoryByteSecondsUsed       uint64            `json:"memoryByteSecondsUsed"`       // @bingen:field[version=1] GPU memory usage in byte-seconds
+	PowerWattSeconds            float64           `json:"powerWattSeconds"`            // @bingen:field[version=1] GPU power consumption in watt-seconds (Joules)
+	PowerWattAverage            float64           `json:"powerWattAverage"`            // @bingen:field[version=1] GPU average power consumption in watts
+	PowerWattMax                float64           `json:"powerWattMax"`                // @bingen:field[version=1] GPU max power consumption in watts
```
should the watts get moved to the device? and then device loses the Usage percentage/max etc?

spa-raj replied:
Moved Power fields to GPUDevice and removed from Usage.

ameijer commented on core/pkg/model/kubemodel/gpuusage.go:
```diff
@@ -0,0 +1,92 @@
+	PowerWattSeconds            float64           `json:"powerWattSeconds"`            // @bingen:field[version=1] GPU power consumption in watt-seconds (Joules)
+	PowerWattAverage            float64           `json:"powerWattAverage"`            // @bingen:field[version=1] GPU average power consumption in watts
+	PowerWattMax                float64           `json:"powerWattMax"`                // @bingen:field[version=1] GPU max power consumption in watts
+	Diagnostic                  *DiagnosticResult `json:"diagnostic,omitempty"`        // @bingen:field[version=1]
```
toss diagnostics since we have that in kubemodel root?

ameijer commented on core/pkg/model/kubemodel/node.go:
```diff
@@ -0,0 +1,26 @@
+	CpuMillicoreUsageAverage uint64            `json:"cpuMillicoreUsageAverage"` // @bingen:field[version=1]
+	CpuMillicoreUsageMax     uint64            `json:"cpuMillicoreUsageMax"`     // @bingen:field[version=1]
+	RAMByteUsageAverage      uint64            `json:"ramByteUsageAverage"`      // @bingen:field[version=1]
+	RAMByteUsageMax          uint64            `json:"ramByteUsageMax"`          // @bingen:field[version=1]
```
I think all these usage fields can be removed. We can get the usage of the node by summing the usage of all the containers on the node. What we care about from the node is how many CPU cores are allocatable, which we get from the CPUMillicoreseconds and etc above

Sean-Holcomb replied:
I don't think that we can calculate maxes or even averages that way. If these are being pulled from metrics that we have access to then I dont think we should remove these fields

ameijer replied:
assuming that we can get this information separately from container level maxes, then we should keep these. if not, then we should remove

spa-raj replied:
I dug into how the existing OpenCost code handles node usage and you're both right.

The current cluster.go Node struct doesn't store simple avg/max values - it uses breakdowns (idle/system/user percentages) from `QueryNodeCPUModeTotal` and similar queries. These give way more insight than just averages.

@Sean-Holcomb - you're right that node max can't be computed from container maxes. But we don't have node-level usage queries in kubemodel yet anyway.

 @ameijer  - your condition applies: we can't get this separately from containers right now, so we should remove these fields.

**Proposing to remove:**
  - `CpuMillicoreUsageAverage`
  - `CpuMillicoreUsageMax`
  - `RAMByteUsageAverage`
  - `RAMByteUsageMax`

If we need node-level usage metrics later, we should implement them properly with breakdowns or node_exporter queries, not half-populated fields. For now, nodes track capacity, containers track usage.

What do you think?

ameijer replied:
yes I think we should go ahead and remove these then since we never collected them in the first place. @Sean-Holcomb should we track a TODO to implement getting these for nodes from somewhere?

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	StorageByteSeconds         uint64            `json:"storageByteSeconds"`         // @bingen:field[version=1]
+	StorageByteRequestAverage  uint64            `json:"storageByteRequestAverage"`  // @bingen:field[version=1]
+	StorageByteUsageAverage    uint64            `json:"storageByteUsageAverage"`    // @bingen:field[version=1]
+	StorageByteUsageMax        uint64            `json:"storageByteUsageMax"`        // @bingen:field[version=1]
```
I think we need to think about putting this in a map of different storage types

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,28 @@
+	RAMByteRequestAverage      uint64            `json:"ramByteRequestAverage"`      // @bingen:field[version=1]
+	RAMByteUsageAverage        uint64            `json:"ramByteUsageAverage"`        // @bingen:field[version=1]
+	RAMByteUsageMax            uint64            `json:"ramByteUsageMax"`            // @bingen:field[version=1]
+	StorageByteSeconds         uint64            `json:"storageByteSeconds"`         // @bingen:field[version=1]
```
I don't think pod needs to duplicate any of the fields in the containers - with the exception of network (which I do think needs to stay since we only have network stats at pod level), all of these CPU*, RAM*, Storage fields can be calculated from the containers under pod

spa-raj replied:
yes, I have removed these duplicate fields from pod.

ameijer commented on core/pkg/model/kubemodel/pod.go:
```diff
@@ -0,0 +1,28 @@
+	StorageByteSeconds         uint64            `json:"storageByteSeconds"`         // @bingen:field[version=1]
+	NetworkTransferBytes       uint64            `json:"networkTransferBytes"`       // @bingen:field[version=1]
+	NetworkReceiveBytes        uint64            `json:"networkReceiveBytes"`        // @bingen:field[version=1]
+	Diagnostic                 *DiagnosticResult `json:"diagnostic,omitempty"`       // @bingen:field[version=1]
```
toss?

ameijer commented on core/pkg/model/kubemodel/pvc.go:
```diff
@@ -0,0 +1,22 @@
+	UID                 string            `json:"uid"`                   // @bingen:field[version=1]
+	NamespaceUID        string            `json:"namespaceUid"`          // @bingen:field[version=1]
+	VolumeUID           *string           `json:"volumeUid,omitempty"`   // @bingen:field[version=1]
+	PodUID              *string           `json:"podUid,omitempty"`      // @bingen:field[version=1]
```
should this be an array? A PVC will frequently be attached to multiple pods in a window if a pod for example moves. I almost wonder if we need like a map here that maps a pod UID to a start,end pair so we can tell when the PVC was attached to which pod in a window

ameijer commented on pkg/costmodel/kubemodel.go:
```diff
@@ -0,0 +1,872 @@
+			Timestamp:   time.Now().UTC(),
+			Error:       err.Error(),
+		}
+		kms.Metadata.Diagnostics = append(kms.Metadata.Diagnostics, diagnostic)
```
looks like you are indeed loading the diagnostics only in the kube model set. let's toss diagnostics from the other objects

spa-raj committed:
Removed Diagnostics from individual structs

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

Sean-Holcomb commented on core/pkg/model/kubemodel/diagnostic.go:
```diff
@@ -0,0 +1,14 @@
+
+// @bingen:generate:DiagnosticResult
+type DiagnosticResult struct {
+	UID         string            `json:"uid"`               // @bingen:field[version=1]
```
What is the UID of a DiagnosticResult?

spa-raj replied:
I was thinking of UUIDv7 like in https://github.com/opencost/opencost/blob/develop/core/pkg/diagnostics/service.go#L148

Sean-Holcomb commented on core/pkg/model/kubemodel/gpudevice.go:
```diff
@@ -0,0 +1,75 @@
+	ModelName                   string  `json:"modelName"`                   // @bingen:field[version=1]
+	IsShared                    bool    `json:"isShared"`                    // @bingen:field[version=1] GPU sharing information
+	SharePercentage             float64 `json:"sharePercentage"`             // @bingen:field[version=1]
+	GpuSeconds                  float64 `json:"gpuSeconds"`                  // @bingen:field[version=1] GPU seconds available
```
flagging this to investigate what the metric populating looks like here. Could we use "MilliGPU"?

spa-raj replied:
I looked into this and MilliGPU doesn't really match how the industry works.

  Here's what I found:

  - Kubernetes: GPUs are allocated as whole integers (nvidia.com/gpu: 1). You can't request fractional GPUs natively.
  - Cloud providers (AWS/GCP/Azure): They bill in whole GPU-hours or GPU-seconds (like "$0.74/GPU/hour"), not milliGPUs.
  - GPU sharing (MIG, time-slicing): When GPUs are shared, they use decimal fractions like 0.25 or 0.5 GPU, not millicores.
  - DCGM metrics: Reports utilization as percentages (0-100%), not millicores.
  - Intel exception: Intel GPUs do use millicores, but that's Intel-specific, not an industry standard.

  I'd suggest keeping GpuSeconds as float64 representing GPU device-seconds. This naturally handles fractional GPU usage:
  - 3600.0 = 1 whole GPU for 1 hour
  - 1800.0 = 1 whole GPU for 30 minutes (or 0.5 GPU for 1 hour)
  - 900.0 = 0.25 GPU for 1 hour

  This matches how cloud providers bill and how time-slicing works. CPUs use millicores because they're built for fine-grained time-sharing, but GPUs are different—they're
  monolithic compute devices where the base unit is still a whole GPU (or GPU instance for MIG).

ameijer commented on core/pkg/model/kubemodel/gpudevice.go:
```diff
@@ -0,0 +1,75 @@
+	DeviceNumber                int32   `json:"deviceNumber"`                // @bingen:field[version=1]
+	ModelName                   string  `json:"modelName"`                   // @bingen:field[version=1]
+	IsShared                    bool    `json:"isShared"`                    // @bingen:field[version=1] GPU sharing information
+	SharePercentage             float64 `json:"sharePercentage"`             // @bingen:field[version=1]
```
GPU can only be mounted to one node at a time

spa-raj replied:
SharePercentage - it's required for cost attribution even though DCGM metrics already scope to virtual GPU slices.

  ---
  How DCGM Metrics Handle Virtual GPUs

  | Aspect                | Behavior                                                        |
  |-----------------------|-----------------------------------------------------------------|
  | UUID labels           | MIG slices get distinct UUIDs (MIG-GPU-<uuid>)                  |
  | Utilization metrics   | DCGM_FI_PROF_GR_ENGINE_ACTIVE reports 0-1 relative to the slice |
  | Per-container scoping | Metrics grouped by (container, pod, namespace, UUID)            |

  A MIG slice at 100% utilization reports 1.0, not 0.143 (1/7th of physical).

  ---
  Why SharePercentage Is Still Needed

  Metrics tell you: "How much of my allocated slice am I using?"
  SharePercentage tells you: "What fraction of the physical GPU is my slice?"

  | Scenario                   | Without SharePercentage | With SharePercentage          |
  |----------------------------|-------------------------|-------------------------------|
  | MIG 1/7 slice @ 100% usage | Billed as "1 GPU" ❌     | Billed as 14.3% of GPU cost ✅ |
  | Time-sliced (4 pods)       | Each sees 100% capacity | Each pays 25% of cost         |

  ---
  Current Implementation Gap

  IsGPUShared:     ✅ Populated (from nvidia_com_gpu_shared resource type)
  SharePercentage: ❌ NOT populated by Prometheus queries

  To populate SharePercentage, we'd need:
  - Node-level GPU inventory data
  - MIG partition configuration
  - Or explicit annotation/label on the workload

  ---
  Recommendation

  Keep both fields in GPUDevice:
  - IsShared - boolean flag for quick filtering
  - SharePercentage - required for accurate cost attribution

  Future work: Add a data source for MIG/vGPU partition configuration to populate SharePercentage.

ameijer commented on core/pkg/model/kubemodel/service.go:
```diff
@@ -0,0 +1,28 @@
+	NamespaceUID         string            `json:"namespaceUid"`          // @bingen:field[version=1]
+	Name                 string            `json:"name"`                  // @bingen:field[version=1]
+	Type                 string            `json:"type"`                  // @bingen:field[version=1]
+	Labels               map[string]string `json:"labels,omitempty"`      // @bingen:field[version=1]
```
add optional field for the hostname

aaliomer-ibm commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,24 @@
+	RAMByteSeconds             uint64            `json:"ramByteSeconds"`             // @bingen:field[version=1]
+	RAMByteRequestAverage      uint64            `json:"ramByteRequestAverage"`      // @bingen:field[version=1]
+	RAMByteUsageAverage        uint64            `json:"ramByteUsageAverage"`        // @bingen:field[version=1]
+	RAMByteUsageMax            uint64            `json:"ramByteUsageMax"`            // @bingen:field[version=1]
```
average is not worth storing. cumulative  to average is a simple calculation (byte-hours to average bytes for example). it makes sense to store max however just not averages

@ameijer @spa-raj  what do you think? I never seen these averages used anywhere

ameijer replied:
I think I agree with you

aaliomer-ibm replied:
to clarify, i consider average a computed value in this scenario. We should never store computed values. 
I'm sure im wrong and there is some context but at least wanted to share my idea

spa-raj replied:
I agree with you. I think we can calculate averages like this
`RAMByteUsageAverage = RAMByteSeconds / (End - Start).Seconds()`

Let's discuss this in our standup.

spa-raj committed:
Add ServiceType enum and Hostname field to Service struct

- Introduces type-safe ServiceType enum (ClusterIP, NodePort, LoadBalancer, ExternalName) and adds optional Hostname field for LoadBalancer service tracking.
- Also marks Diagnostics as omitempty in Metadata.

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Move GPU power metrics from GPUUsage to GPUDevice

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
- Remove redundant Pod CPU/RAM/Storage fields aggregated from Containers
- Network metrics retained as pod-level only since containers share network namespace

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Retain Pod max usage fields as they cannot be aggregated from containers

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

ameijer committed:
Merge branch 'develop' into feature/kubemodel-flat

ameijer commented on core/pkg/model/kubemodel/container.go:
```diff
@@ -0,0 +1,23 @@
+	Name                       string    `json:"name"`                       // @bingen:field[version=1]
+	Start                      time.Time `json:"start"`                      // @bingen:field[version=1]
+	End                        time.Time `json:"end"`                        // @bingen:field[version=1]
+	CpuMillicoreSeconds        uint64    `json:"cpuMillicoreSeconds"`        // @bingen:field[version=1]
```
```suggestion
	CpuMillicoreSecondsAllocated        uint64    `json:"cpuMillicoreSeconds"`        // @bingen:field[version=1]
```
Same for the others

ameijer commented:
goald before thanksgiving:
 merge everything EXCEPT 
1. GPUs (removed from kubemodel set etc)
2. Volumes (removed from kubemodel set etc)

ameijer commented on core/pkg/model/kubemodel/controller.go:
```diff
@@ -0,0 +1,27 @@
+// @bingen:generate:Controller
+type Controller struct {
+	UID          string            `json:"uid"`                   // @bingen:field[version=1]
+	NamespaceUID string            `json:"namespaceUid"`          // @bingen:field[version=1]
```
OwnerID as well

ameijer commented on core/pkg/model/kubemodel/controller.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+// @bingen:generate:Controller
+type Controller struct {
```
change to Owner

spa-raj committed:
Rename resource allocation fields to include 'Allocated' suffix across kubemodel

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

spa-raj committed:
Refactor kubemodel: rename Controller to Owner with ControllerUID to OwnerUID and fix struct spacing

Signed-off-by: Sparsh <sparsh.raj30@gmail.com>

Sean-Holcomb commented:
closing in favor these changes to KubeModel https://github.com/opencost/opencost/pull/3875
